segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -392,66 +401,213 @@ private boolean isFatalException(Throwable ex) {
+            try {
+                // Record the end of a frame in the DurableDataLog. No need for locking here, as the metadata has its own.
+                this.metadataUpdater.recordTruncationMarker(commitArgs.getLastStartedSequenceNumber(), commitArgs.getLogAddress());
+                final long lastOperationSequence = commitArgs.getLastFullySerializedSequenceNumber();
+                final long addressSequence = commitArgs.getLogAddress().getSequence();
+
+                synchronized (this.lock) {
+                    // Record the Truncation marker. The OperationMetadataUpdater will update the Metadata directly for this.
+                    if (addressSequence <= this.highestCommittedDataFrame) {
+                        // Ack came out of order (we already processed one with a higher SeqNo).
+                        log.debug("{}: CommitRejected ({}, HighestCommittedDataFrame = ).", this.traceObjectId, commitArgs, this.highestCommittedDataFrame);
+                        return;
+                    }
+
+                    // Commit any changes to the metadata.
+                    boolean checkpointExists = this.metadataTransactions.removeFirst(commitArgs);
+                    assert checkpointExists : "No Metadata UpdateTransaction found for " + commitArgs;
+                    this.metadataUpdater.commit(commitArgs.key());
+
+                    // Acknowledge all pending entries, in the order in which they are in the queue (ascending seq no).
+                    while (!this.pendingOperations.isEmpty()
+                            && this.pendingOperations.peekFirst().getOperation().getSequenceNumber() <= lastOperationSequence) {
+                        CompletableOperation op = this.pendingOperations.pollFirst();
+                        try {
+                            this.logUpdater.process(op.getOperation());
+                        } catch (Throwable ex) {
+                            // MemoryStateUpdater.process() should only throw DataCorruptionExceptions, but just in case it
+                            // throws something else (i.e. NullPtr), we still need to handle it.
+                            // First, fail the operation, since it has already been taken off the pending list.
+                            log.error("{}: OperationCommitFailure ({}). {}", this.traceObjectId, op.getOperation(), ex);
+                            toFail.put(op, ex);
+
+                            // Then fail the remaining operations (which also handles fatal errors) and bail out.
+                            collectFailureCandidates(ex, commitArgs, toFail);
+                            if (isFatalException(ex)) {
+                                this.stopException = ex;
+                                CallbackHelpers.invokeSafely(this.fatalExceptionCallback, ex, null);
+                            }
+
+                            return;
+                        }
+
+                        if (this.stopException != null) {
```

Comment:
This is the callback for a successful DataFrame commit. The line you're pointing at is reached only if the operation itself caused a fatal exception. Regardless of the operation status, it is still correct to update `highestCommittedDataFrame` since that is an accurate representation of the Tier1 state.